### ISSUE TITLE
strange errors: lws_ssl_capable_read failed w/ DH lib and w/ EVP lib

### ISSUE THREAD

**Issue by namowen** [warmcat/libwebsockets#802]
I just cloned master and run test-server and test-client.
1. error, alway repeating:
lwsts[125413]: lws_ssl_capable_read failed: error:00000005:lib(0):func(0):DH lib
2 error in server log when client exits (sometimes shows once, sometimes twice):
lwsts[125413]: lws_ssl_capable_read failed: error:00000006:lib(0):func(0):EVP lib

Client started w/: libwebsockets-test-client --ssl localhost
Here the log of the server, started w/ libwebsockets-test-server --ssl

lwsts[125413]: libwebsockets test server - license LGPL2.1+SLE
lwsts[125413]: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
Using resource path "/usr/local/share/libwebsockets-test-server"
lwsts[125413]: Initial logging level 7
lwsts[125413]: Libwebsockets version: 2.1.0 bzuber@iotalive1-v2.0.0-270-gc7c4ae0
lwsts[125413]: IPV6 compiled in and enabled
lwsts[125413]: libev support not compiled in
lwsts[125413]: libuv support not compiled in
lwsts[125413]:  Threads: 1 each 1024 fds
lwsts[125413]:  mem: platform fd map:  8192 bytes
lwsts[125413]:  Compiled with OpenSSL support
lwsts[125413]: Doing SSL library init
lwsts[125413]: Creating Vhost 'default' port 7681, 5 protocols, IPv6 on
lwsts[125413]:  SSL ciphers: 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!MD5:!PSK:!RC4:!HMAC_SHA1:!SHA1:!DHE-RSA-AES128-GCM-SHA256:!DHE-RSA-AES128-SHA256:!AES128-GCM-SHA25
lwsts[125413]:  Using SSL mode
lwsts[125413]:  SSL ECDH curve 'prime256v1'
lwsts[125413]:  Listening on port 7681
lwsts[125413]:  mem: per-conn:          568 bytes + protocol rx buf
lwsts[125413]:  canonical_hostname = iotalive1
lwsts[125413]: SNI: Unknown ServerName: localhost
lwsts[125413]: SNI: Unknown ServerName: localhost
    get  = /
    host: = localhost
    connection: = Upgrade
    upgrade: = websocket
    origin: = http://localhost
    sec-websocket-extensions: = permessage-deflate; client_max_window_bits
    sec-websocket-protocol: = dumb-increment-protocol
    http/1.1  = HTTP/1.1
    pragma: = no-cache
    cache-control: = no-cache
    sec-websocket-key: = ro/P7p8IWFRb+VbL27QUqw==
    sec-websocket-version: = 13
lwsts[125413]:  permessage-deflate requires the protocol (dumb-increment-protocol) to have an RX buffer >= 128
lwsts[125413]: ext permessage-deflate failed construction
    get  = /
    host: = localhost
    connection: = Upgrade
    upgrade: = websocket
    origin: = http://localhost
    sec-websocket-extensions: = permessage-deflate; client_max_window_bits
    sec-websocket-protocol: = lws-mirror-protocol
    http/1.1  = HTTP/1.1
    pragma: = no-cache
    cache-control: = no-cache
    sec-websocket-key: = vm3UvsBNiKWUCTux/AUe3Q==
    sec-websocket-version: = 13
lwsts[125413]:  Capping pmd rx to 128
lwsts[125413]: lws_ssl_capable_read failed: error:00000005:lib(0):func(0):DH lib
lwsts[125413]: SNI: Unknown ServerName: localhost
    get  = /
    host: = localhost
    connection: = Upgrade
    upgrade: = websocket
    origin: = http://localhost
    sec-websocket-extensions: = permessage-deflate; client_max_window_bits
    sec-websocket-protocol: = lws-mirror-protocol
    http/1.1  = HTTP/1.1
    pragma: = no-cache
    cache-control: = no-cache
    sec-websocket-key: = 2G+G6NTEBH7QVPS4D3U2UA==
    sec-websocket-version: = 13
lwsts[125413]:  Capping pmd rx to 128
lwsts[125413]: lws_ssl_capable_read failed: error:00000005:lib(0):func(0):DH lib
lwsts[125413]: SNI: Unknown ServerName: localhost
    get  = /
    host: = localhost
    connection: = Upgrade
    upgrade: = websocket
    origin: = http://localhost
    sec-websocket-extensions: = permessage-deflate; client_max_window_bits
    sec-websocket-protocol: = lws-mirror-protocol
    http/1.1  = HTTP/1.1
    pragma: = no-cache
    cache-control: = no-cache
    sec-websocket-key: = tXna55Bgr42O/NSjK6CqXA==
    sec-websocket-version: = 13
lwsts[125413]:  Capping pmd rx to 128
lwsts[125413]: lws_ssl_capable_read failed: error:00000006:lib(0):func(0):EVP lib
lwsts[125413]: lws_ssl_capable_read failed: error:00000006:lib(0):func(0):EVP lib
^Clwsts[125413]: lws_context_destroy: ctx 0x6c5010
lwsts[125413]: callback_lws_mirror: mirror protocol cleaning up
lwsts[125413]: lws_context_destroy2: ctx 0x6c5010
lwsts[125413]: libwebsockets-test-server exited cleanly

##################################################
client log:

[2017/02/18 13:12:38:7130] NOTICE: libwebsockets test client - license LGPL2.1+SLE
[2017/02/18 13:12:38:7130] NOTICE: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
[2017/02/18 13:12:38:7130] NOTICE:  Using SSL
[2017/02/18 13:12:38:7130] NOTICE:  Selfsigned certs allowed
[2017/02/18 13:12:38:7131] NOTICE:  Skipping peer cert hostname check
[2017/02/18 13:12:38:7131] NOTICE: Initial logging level 7
[2017/02/18 13:12:38:7131] NOTICE: Libwebsockets version: 2.1.0 bzuber@iotalive1-v2.0.0-270-gc7c4ae0
[2017/02/18 13:12:38:7131] NOTICE: IPV6 compiled in and enabled
[2017/02/18 13:12:38:7131] NOTICE: libev support not compiled in
[2017/02/18 13:12:38:7131] NOTICE: libuv support not compiled in
[2017/02/18 13:12:38:7131] NOTICE:  Threads: 1 each 1024 fds
[2017/02/18 13:12:38:7131] NOTICE:  mem: platform fd map:  8192 bytes
[2017/02/18 13:12:38:7131] NOTICE:  Compiled with OpenSSL support
[2017/02/18 13:12:38:7131] NOTICE: Doing SSL library init
[2017/02/18 13:12:38:7146] NOTICE: Creating Vhost 'default' port -1, 2 protocols, IPv6 on
[2017/02/18 13:12:38:7149] NOTICE:  mem: per-conn:          568 bytes + protocol rx buf
[2017/02/18 13:12:38:7149] NOTICE:  canonical_hostname = iotalive1
[2017/02/18 13:12:38:7149] NOTICE: using  mode (ws)
[2017/02/18 13:12:38:7149] NOTICE: dumb: connecting
[2017/02/18 13:12:38:7149] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:38:7155] NOTICE: mirror: connecting
[2017/02/18 13:12:38:7156] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:38:7158] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:38:7170] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:38:7181] NOTICE: accepting self-signed certificate (verify_callback)
[2017/02/18 13:12:38:7190] NOTICE: lws_ssl_client_connect2: SSL_connect says -1
[2017/02/18 13:12:38:7192] NOTICE: accepting self-signed certificate (verify_callback)
[2017/02/18 13:12:38:7199] NOTICE: lws_ssl_client_connect2: SSL_connect says -1
[2017/02/18 13:12:38:7201] NOTICE: lws_ssl_client_connect2: SSL_connect says 1
[2017/02/18 13:12:38:7206] NOTICE: lws_ssl_client_connect2: SSL_connect says 1
[2017/02/18 13:12:38:7220] NOTICE: checking client ext permessage-deflate
[2017/02/18 13:12:38:7220] NOTICE: instantiating client ext permessage-deflate
[2017/02/18 13:12:38:7220] ERR:  Capping pmd rx to 128
[2017/02/18 13:12:38:7220] NOTICE: mirror: LWS_CALLBACK_CLIENT_ESTABLISHED
[2017/02/18 13:12:41:5836] NOTICE: mirror: LWS_CALLBACK_CLOSED mirror_lifetime=0
[2017/02/18 13:12:41:5837] NOTICE: mirror: connecting
[2017/02/18 13:12:41:5837] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:41:5839] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:41:5849] NOTICE: accepting self-signed certificate (verify_callback)
[2017/02/18 13:12:41:5862] NOTICE: checking client ext permessage-deflate
[2017/02/18 13:12:41:5863] NOTICE: instantiating client ext permessage-deflate
[2017/02/18 13:12:41:5863] ERR:  Capping pmd rx to 128
[2017/02/18 13:12:41:5863] NOTICE: mirror: LWS_CALLBACK_CLIENT_ESTABLISHED
[2017/02/18 13:12:45:1181] NOTICE: mirror: LWS_CALLBACK_CLOSED mirror_lifetime=0
[2017/02/18 13:12:45:1183] NOTICE: mirror: connecting
[2017/02/18 13:12:45:1183] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:45:1186] NOTICE: lws_client_connect_2: address localhost
[2017/02/18 13:12:45:1196] NOTICE: accepting self-signed certificate (verify_callback)
[2017/02/18 13:12:45:1201] NOTICE: lws_ssl_client_connect2: SSL_connect says -1
[2017/02/18 13:12:45:1205] NOTICE: lws_ssl_client_connect2: SSL_connect says 1
[2017/02/18 13:12:45:1221] NOTICE: checking client ext permessage-deflate
[2017/02/18 13:12:45:1221] NOTICE: instantiating client ext permessage-deflate
[2017/02/18 13:12:45:1221] ERR:  Capping pmd rx to 128
[2017/02/18 13:12:45:1221] NOTICE: mirror: LWS_CALLBACK_CLIENT_ESTABLISHED
^C[2017/02/18 13:12:45:1360] ERR: Exiting
[2017/02/18 13:12:45:1360] NOTICE: lws_context_destroy: ctx 0x1e1f760
[2017/02/18 13:12:45:1360] NOTICE: dumb: LWS_CALLBACK_CLOSED
[2017/02/18 13:12:45:1361] NOTICE: mirror: LWS_CALLBACK_CLOSED mirror_lifetime=37266
[2017/02/18 13:12:45:1363] NOTICE: lws_context_destroy2: ctx 0x1e1f760

**Comment by lws-team** [comment]
Yeah I can reproduce it.  But they seem to occur once per connection and the actual business of spamming circles on the shared canvas seems to proceed smoothly.... is that what you are seeing?

**Comment by namowen** [comment]
EVP lib error happens if the server gets closed (shutdown) and client still runs. It looks like if one of the peers (server or clients) stops, the other peers issues the EVP lib error.

[2017/02/18 16:00:15:6476] NOTICE: mirror: LWS_CALLBACK_CLIENT_ESTABLISHED
[2017/02/18 16:00:16:9867] ERR: lws_ssl_capable_read failed: error:00000006:lib(0):func(0):EVP lib
[2017/02/18 16:00:16:9868] NOTICE: mirror: LWS_CALLBACK_CLOSED mirror_lifetime=52293
[2017/02/18 16:00:16:9871] ERR: lws_ssl_capable_read failed: error:00000006:lib(0):func(0):EVP lib
[2017/02/18 16:00:16:9872] NOTICE: dumb: LWS_CALLBACK_CLOSED

And yes, the circles are drawn, even when the DH lib error occures.

BTW I use OpenSSL 1.0.2j

**Comment by namowen** [comment]
I tried following w/o any help:
- selfsigned ec cert instead of selfsigned rsa cert
- limit ciphers list to ecdhe only
- openssl 1.0.2h (instead of openssl 1.0.2.j)
- removed OpenSSL_client_verify_callback in ssl-client.c

**Comment by lws-team** [comment]
As you say this error is only coming when we kill the connection... I think it's probably nothing new, just that master reports the SSL error stack in more places than before.  The test client tests repeated connections by giving each one a random "lifetime" and then closing it and opening a new one.  So the test client shows this at intervals repeatedly.

I agree it's a bit scary showing it as ERR... not sure what to do about it.

**Comment by namowen** [comment]
Here a patch giving some light into the problem with printing out some info on lwsl_err level
It seems that SSL_shutdown does not behave as expected.
EVP lib error is definitely a received SSL_RECEIVED_SHUTDOWN in SSL_Read, so it should be handled accordingly (closing the SSL session) and not writing out EVP lib error anymore.
The DH lib error seems to be more tricky. The log shows, that the client is not successful in SSL_shutdown (returning -1), and the server is no longer capable to SSL_read (produces DH lib error in server). I assume the client is the trigger of the close, not the server.
It looks like the client has no connection to the server when it starts its SSL_Shutdown. This must be investigated further.
Client log (part):
[2017/02/19 19:13:40:0469] NOTICE: mirror: LWS_CALLBACK_CLIENT_ESTABLISHED
[2017/02/19 19:13:42:2409] NOTICE: mirror: LWS_CALLBACK_CLOSED mirror_lifetime=0
[2017/02/19 19:13:42:2411] ERR: lws_ssl_close: SSL_shutdown=-1
[2017/02/19 19:13:42:2412] ERR: lws_ssl_close failed: error:00000005:lib(0):func(0):DH lib
[2017/02/19 19:13:42:2413] NOTICE: mirror: connecting
Server log (part):
lwsts[125580]: SSL_get_shutdown: 0
lwsts[125580]: SSL_get_shutdown: No shutdown setting, yet.
lwsts[125580]: lws_ssl_capable_read failed0: error:00000005:lib(0):func(0):DH lib
lwsts[125580]: lws_ssl_close: SSL_shutdown=-1
lwsts[125580]: lws_ssl_close failed: error:00000005:lib(0):func(0):DH lib

Patch for printouts:
######################################
diff --git a/lib/ssl.c b/lib/ssl.c
index a616671..8d5e0ab 100644
--- a/lib/ssl.c
+++ b/lib/ssl.c
@@ -336,10 +336,28 @@
 
 	/* manpage: returning 0 means connection shut down */
 	if (!n) {
+		/* int SSL_get_shutdown(const SSL *ssl); */
+		/* # define SSL_SENT_SHUTDOWN       1
+		   # define SSL_RECEIVED_SHUTDOWN   2
+		 */
+		int shutdown_state = SSL_get_shutdown(wsi->ssl);
+		lwsl_err("SSL_get_shutdown: %d\n", shutdown_state);
+		if (shutdown_state & SSL_RECEIVED_SHUTDOWN) {
+			lwsl_err("SSL_get_shutdown: SSL_RECEIVED_SHUTDOWN\n");
 			lwsl_err("%s failed: %s\n",__func__,
-				 ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
+				 ERR_error_string(lws_ssl_get_error(wsi, 0 /*n*/), NULL));
+		} else if(shutdown_state == 0) {
+			lwsl_err("SSL_get_shutdown: No shutdown setting, yet.\n");
+			lwsl_err("%s failed0: %s\n",__func__,
+				 ERR_error_string(lws_ssl_get_error(wsi, 0 /*n*/), NULL));
 			lws_decode_ssl_error();
-
+		} else if(shutdown_state  & SSL_SENT_SHUTDOWN) {
+			lwsl_err("SSL_get_shutdown: SSL_SENT_SHUTDOWN\n");
+		} else {
+			lwsl_err("%s failed: %s\n",__func__,
+				 ERR_error_string(lws_ssl_get_error(wsi, 0 /*n*/), NULL));
+			lws_decode_ssl_error();
+		}
 		return LWS_SSL_CAPABLE_ERROR;
 	}
 
@@ -468,7 +486,12 @@
 #if defined(LWS_USE_MBEDTLS)
 #else
 	n = SSL_get_fd(wsi->ssl);
-	SSL_shutdown(wsi->ssl);
+	int ret = SSL_shutdown(wsi->ssl);
+	lwsl_err("%s: SSL_shutdown=%d\n",__func__, ret);
+	if (ret == 0) lwsl_err("%s: SSL_shutdown notify alert sent successfully, awaiting response from peer\n",__func__);
+	if (ret == 1) lwsl_err("%s: 2nd SSL_shutdown notify alert sent successfully, after receiving notify alert response from peer\n",__func__);
+	if (ret < 0) lwsl_err("%s failed: %s\n",__func__,
+		 ERR_error_string(lws_ssl_get_error(wsi, ret), NULL));
 	compatible_close(n);
 	SSL_free(wsi->ssl);
 #endif
@@ -641,7 +664,13 @@
 #else
 #if defined(LWS_USE_MBEDTLS)
 #else
-				SSL_shutdown(wsi->ssl);
+				int ret = SSL_shutdown(wsi->ssl);
+				lwsl_err("%s: SSL_shutdown=%d\n",__func__, ret);
+				if (ret == 0) lwsl_err("%s: SSL_shutdown notify alert sent successfully, awaiting response from peer\n",__func__);
+				if (ret == 1) lwsl_err("%s: 2nd SSL_shutdown notify alert sent successfully, after receiving notify alert response from peer\n",__func__);
+				if (ret < 0) lwsl_err("%s failed: %s\n",__func__,
+					 ERR_error_string(lws_ssl_get_error(wsi, ret), NULL));
+
 				SSL_free(wsi->ssl);
 #endif
 #endif

**Comment by namowen** [comment]
testclient w/ -d15 shows: connection is teared down before SSL_shutdown is made.
This must be fixed ;-)
I leave that to you, you know your code better than I do ;-)

[2017/02/19 19:50:46:8035] INFO: lws_read: read_ok, used 19
[2017/02/19 19:50:46:8036] INFO: closing mirror session
[2017/02/19 19:50:46:8036] INFO: lws_service_fd: closing
[2017/02/19 19:50:46:8036] INFO: lws_close_free_wsi: shutting down connection: 0x74ffe0 (sock 7, state 5)
[2017/02/19 19:50:46:8037] INFO: lws_read: read_ok, used 20
[2017/02/19 19:50:46:8037] INFO: lws_read: read_ok, used 18
[2017/02/19 19:50:46:8037] INFO: lws_read: read_ok, used 19
[2017/02/19 19:50:46:8053] INFO: lws_close_free_wsi: real just_kill_connection: 0x74ffe0 (sockfd 7)
[2017/02/19 19:50:46:8053] INFO: remove_wsi_socket_from_fds: removing same prot wsi 0x74ffe0
[2017/02/19 19:50:46:8053] INFO: have prev 0x74c7c8, setting him to our next (nil)
[2017/02/19 19:50:46:8053] NOTICE: mirror: LWS_CALLBACK_CLOSED mirror_lifetime=0
[2017/02/19 19:50:46:8055] ERR: lws_ssl_close: SSL_shutdown=-1
[2017/02/19 19:50:46:8055] ERR: lws_ssl_close failed: error:00000005:lib(0):func(0):DH lib
[2017/02/19 19:50:46:8056] INFO: ah det due to close
[2017/02/19 19:50:46:8056] INFO: lws_header_table_detach: wsi 0x74ffe0: ah (nil) (tsi=0, count = 0)
[2017/02/19 19:50:46:8056] INFO: lws_header_table_detach: wsi 0x74ffe0: ah (nil) (tsi=0, count = 0)
[2017/02/19 19:50:46:8056] NOTICE: mirror: connecting

**Comment by lws-team** [comment]
"must" is a pretty strong word you seldom see used unqualified... like someone else's 'need' usually means 'want', someone else's 'must' usually means 'should'.  Lws isn't a job for me, unless I'm using it for something, nothing about it is mandatory to me.

I can see with your explanation (thanks) the problem is we always shutdown the underlying tcp socket in the staged close.  I pushed a patch on master that has it call SSL_shutdown() instead if the wsi was in ssl mode; this will be called again when we finally close but it's expected according to the manpage.

That gets rid of the dh lib trace but not the evp one....  shutdown is saying we're not going to write any more on the connection, it's possible to authoritively say that for non-ssl, but for ssl he may want to write internally even when reading, closing etc.  The evp seems to be coming after we shutdown the connection, on receipt of something that later causes ssl to write an ack but the connection is already shut down.  I dunno if that means we do the close ordering in an undesirable way still for ssl.

**Comment by namowen** [comment]
Sorry about the 'must', but 1. I meant someone has to fix it (whoever), and in the second line I assumed that you know the code better than I. Especially the DH lib case, because it seems to me more complicated than the EVP lib. And 2. I closed the sentence with ;-). But 'should' should/would (?) have been the right word. Sorry, again.
I appreciate your work very much and use lws since 2013 when I found it. Thank you for your fantastic work.

About the EVP lib: the documentation states, if the return value of SSL_read is 0 then:
> The read operation was not successful. The reason may either be a clean shutdown due to a "close notify" alert sent by the peer (in which case the SSL_RECEIVED_SHUTDOWN flag in the ssl shutdown state is set (see SSL_shutdown(3), SSL_set_shutdown(3)). It is also possible, that the peer simply shut down the underlying transport and the shutdown is incomplete. Call SSL_get_error() with the return value ret to find out, whether an error occurred or the connection was shut down cleanly (SSL_ERROR_ZERO_RETURN).
[https://wiki.openssl.org/index.php/Manual:SSL_read(3)](url)

My understanding is that the code should check in case of 0 return from SSL_read:
1. if SSL_RECEIVED_SHUTDOWN was received (SSL_get_shutdown). If yes everthing is ok and the connection can be closed, what lws is already doing, as I understand.
2. if SSL_get_error() returns a SSL_ERROR_ZERO_RETURN. If yes everthing is ok and the connection can be closed, what lws is already doing, as I understand.
3. the code as it now exist: an error occured and should be reported (what lws actually is doing).

Note: ssl.h: # define SSL_ERROR_ZERO_RETURN           **6**
server error: lwsts[126138]: lws_ssl_capable_read failed: error:**00000006**:lib(0):func(0):EVP lib

As my printout patch shows, the EVP lib case is a SSL_RECEIVED_SHUTDOWN. And I assume SSL_get_error() returns a SSL_ERROR_ZERO_RETURN (=6), but lws prints the error string for 6 which is the EVP lib error string (thats my assumption).
Maybe checking return of SSL_get_error() should be enough to handle this case properly.

(after writing all this, I wondering why I have not fixed it in the mean time and send you a patch. But now the case is well documented... ;-)

**Comment by lws-team** [comment]
No worries, I understand.

But I am travelling at the moment and after thismorning I will be blocked on other stuff for a bit.

It'd be great if you could look at EVP if you understand what why that's happening and try to come up a patch.  Maybe it's really about shutdown ordering for SSL if so and it gets too specific let me know what to look at in lws I will try to advise or modify.  If not in a couple of days there will be time again on my side.

**Comment by namowen** [comment]
I will take care, but I can't promise an end date before sunday evening (UTC+1), 26-Feb-2017, because I'm busy, too ;-)
After applying your patch, the situation looks a little bit different, but I think it's still the same EVP lib.

**Comment by lws-team** [comment]
Sure, no worries.  I appreciate any help.

**Comment by namowen** [comment]
Here the patch for master Libwebsockets version: 2.1.0 bzuber@iotalive1-v2.0.0-287-g2e1dcc5. 
It took more time as expected. The whole SSL_shutdown process w/ non-blocking can be quite tricky, especially having SSL_ERROR_WANT_WRITE as result of SSL_shutdown. But you will see ;-). But it's solved.
Now SSL_read (and SSL_write) properly handle the outcome n==0 and don't write any DH lib or EVP lib errors anymore.
I introduced a new function lws_ssl_shutdown in place of SSL_shutdown. As said, a SSL_shutdown can get a SSL_ERROR_WANT_WRITE with errno=Resource temporarily unavailable (11). Sometimes this happened 20% of the time in my tests (test-client/test-server). I solved it with a while retry 5 times w/ a sleep of a second. In my tests when a SSL_ERROR_WANT_WRITE occured, after 2 seconds the SSL_shutdown was successful. Without the while, it would go in a TIMEDOUT WAITING, which can take ~20 seconds. Because I don't know how to handle the SSL_ERROR_WANT_WRITE and prevent the following TIMEDOUT WAITING, I used the while which seems to me an acceptable solution for this case. But maybe you have a better solution.

Here the 4 patch files:
```
diff --git a/lib/context.c b/lib/context.c
index d35b469..9518a80 100644
--- a/lib/context.c
+++ b/lib/context.c
@@ -924,7 +924,19 @@
 
 	while (m--) {
 		pt = &context->pt[m];
+#ifdef LWS_OPENSSL_SUPPORT
+		/* first send 'close notify' alert to all open wsi */
+		for (n = 0; (unsigned int)n < context->pt[m].fds_count; n++) {
+			struct lws *wsi = wsi_from_fd(context, pt->fds[n].fd);
+			if (!wsi)
+				continue;
 
+			if (lws_is_ssl(wsi) && wsi->ssl) {
+				lws_ssl_shutdown(wsi);
+			}
+		}
+#endif
+		/* proper close of wsi */
 		for (n = 0; (unsigned int)n < context->pt[m].fds_count; n++) {
 			struct lws *wsi = wsi_from_fd(context, pt->fds[n].fd);
 			if (!wsi)
```
###########################################################
```
diff --git a/lib/libwebsockets.c b/lib/libwebsockets.c
index 7300e90..fca826b 100755
--- a/lib/libwebsockets.c
+++ b/lib/libwebsockets.c
@@ -436,22 +436,17 @@
 	    reason != LWS_CLOSE_STATUS_NOSTATUS_CONTEXT_DESTROY &&
 	    !wsi->socket_is_permanently_unusable) {
 #ifdef LWS_OPENSSL_SUPPORT
-		if (lws_is_ssl(wsi) && wsi->ssl)
-               {
-                       lwsl_info("%s: shutting down SSL connection: %p (ssl %p, sock %d, state %d)\n", __func__, wsi, wsi->ssl, (int)(long)wsi->sock, wsi->state);
-			n = SSL_shutdown(wsi->ssl);
-                       if (n == 0) /* Complete bidirectional SSL shutdown */
-                               n = SSL_shutdown(wsi->ssl);
-                       n = shutdown(wsi->sock, SHUT_WR);
-               }
+		if (lws_is_ssl(wsi) && wsi->ssl) {
+			lws_ssl_shutdown(wsi);
+		}
 		else
 #endif
 		{
 			lwsl_info("%s: shutting down connection: %p (sock %d, state %d)\n", __func__, wsi, (int)(long)wsi->sock, wsi->state);
 			n = shutdown(wsi->sock, SHUT_WR);
+			if (n)
+				lwsl_debug("closing: shutdown (state %d) ret %d\n", wsi->state, LWS_ERRNO);
 		}
-		if (n)
-			lwsl_debug("closing: shutdown (state %d) ret %d\n", wsi->state, LWS_ERRNO);
 
 // This causes problems with disconnection when the events are half closing connection
 // FD_READ | FD_CLOSE (33)
```
###############################################
```
diff --git a/lib/ssl.c b/lib/ssl.c
index 8c0d92b..ec009fc 100644
--- a/lib/ssl.c
+++ b/lib/ssl.c
@@ -86,6 +86,58 @@
 	}
 }
 
+#ifdef LWS_OPENSSL_SUPPORT
+
+#ifdef _DEBUG
+#if defined(LWS_WITH_NO_LOGS)
+#define print_ssl_shutdown_state1(a,b,c)
+#define print_ssl_shutdown_state2(a,b)
+#define print_ssl_shutdown_outcome(a,b,c,d)
+#else
+#define print_ssl_shutdown_state1(a,b,c) _print_ssl_shutdown_state1(a,b,c)
+#define print_ssl_shutdown_state2(a,b) _print_ssl_shutdown_state2(a,b)
+#define print_ssl_shutdown_outcome(a,b,c,d) _print_ssl_shutdown_outcome(a,b,c,d)
+#endif
+#endif
+
+#ifdef _DEBUG
+
+void _print_ssl_shutdown_state1(int shutdown_state, const struct lws *wsi, const char funcName[]) {
+	if (shutdown_state & SSL_RECEIVED_SHUTDOWN) {
+		if(shutdown_state & SSL_SENT_SHUTDOWN) {
+			lwsl_info("%s SSL_get_shutdown (%d): SSL_RECEIVED_SHUTDOWN & SSL_SENT_SHUTDOWN (wsi=%p)\n",funcName, shutdown_state, wsi);
+		}
+		else lwsl_info("%s SSL_get_shutdown (%d): SSL_RECEIVED_SHUTDOWN (wsi=%p)\n",funcName, shutdown_state, wsi);
+	} else if(shutdown_state == 0) {
+		lwsl_info("%s SSL_get_shutdown (%d): No shutdown setting, yet. (wsi=%p)\n",funcName, shutdown_state, wsi);
+	} else if(shutdown_state  & SSL_SENT_SHUTDOWN) {
+		lwsl_info("%s SSL_get_shutdown (%d): SSL_SENT_SHUTDOWN (wsi=%p)\n",funcName, shutdown_state, wsi);
+	}
+}
+
+void _print_ssl_shutdown_state2(const struct lws *wsi, const char funcName[]) {
+	int shutdown_state = SSL_get_shutdown(wsi->ssl);
+	_print_ssl_shutdown_state1(shutdown_state, wsi, funcName);
+}
+
+void _print_ssl_shutdown_outcome(int outcome, int shutdown_errno, const struct lws *wsi, const char funcName[]) {
+	if (outcome == 0) lwsl_info("%s: SSL_shutdown (%d) notify alert sent successfully, awaiting response from peer (wsi=%p)\n",funcName, outcome, wsi);
+	if (outcome == 1) lwsl_info("%s: 2nd SSL_shutdown (%d) notify alert sent successfully, after receiving notify alert response from peer (wsi=%p)\n",funcName, outcome, wsi);
+	if (outcome < 0) {
+		lwsl_info("%s: SSL_shutdown=%d error (wsi=%p)\n",funcName, outcome, wsi);
+		int sslErr = SSL_get_error(wsi->ssl, outcome);
+		if (sslErr > SSL_ERROR_NONE && sslErr < SSL_ERROR_WANT_ACCEPT) {
+			char *sslErrStr[] = {"SSL_ERROR_NONE","SSL_ERROR_SSL","SSL_ERROR_WANT_READ","SSL_ERROR_WANT_WRITE","SSL_ERROR_WANT_X509_LOOKUP","SSL_ERROR_SYSCALL","SSL_ERROR_ZERO_RETURN","SSL_ERROR_WANT_CONNECT","SSL_ERROR_WANT_ACCEPT"};
+			lwsl_info("%s: SSL_shutdown: %s (%d); errno=%s (%d) (wsi=%p)\n",funcName, sslErrStr[sslErr], sslErr, strerror(shutdown_errno), shutdown_errno, wsi);
+		} else {
+			lwsl_err("%s failed: %s\n",funcName,	ERR_error_string(sslErr, NULL));
+			lws_decode_ssl_error();
+		}
+	}
+}
+#endif
+#endif
+
 static int
 lws_context_init_ssl_pem_passwd_cb(char * buf, int size, int rwflag, void *userdata)
 {
@@ -237,12 +289,34 @@
 	n = SSL_read(wsi->ssl, buf, len);
 
 	/* manpage: returning 0 means connection shut down */
-	if (!n) {
-		lwsl_err("%s failed: %s\n",__func__,
-			 ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
-		lws_decode_ssl_error();
-
+	if (n == 0) {
+#ifndef _DEBUG
+		int err = SSL_get_error(wsi->ssl, n);
+		if (err > SSL_ERROR_NONE && err < SSL_ERROR_WANT_ACCEPT) {
+			return LWS_SSL_CAPABLE_ERROR;
+		} else {
+			lwsl_err("%s failed: %s\n",__func__,
+					ERR_error_string(err, NULL));
+			lws_decode_ssl_error();
+		}
 		return LWS_SSL_CAPABLE_ERROR;
+#else
+		int errno_ssl_read = errno;
+		int err = SSL_get_error(wsi->ssl, n);
+		if (err == SSL_ERROR_ZERO_RETURN) {
+			print_ssl_shutdown_state2(wsi, __func__);
+			lwsl_info("%s: SSL_read: SSL_ERROR_ZERO_RETURN (%d); (wsi=%p)\n",__func__, SSL_ERROR_ZERO_RETURN, wsi);
+			return LWS_SSL_CAPABLE_ERROR;
+		} else if (err > SSL_ERROR_NONE && err < SSL_ERROR_WANT_ACCEPT) {
+			char *sslErrStr[] = {"SSL_ERROR_NONE","SSL_ERROR_SSL","SSL_ERROR_WANT_READ","SSL_ERROR_WANT_WRITE","SSL_ERROR_WANT_X509_LOOKUP","SSL_ERROR_SYSCALL","SSL_ERROR_ZERO_RETURN","SSL_ERROR_WANT_CONNECT","SSL_ERROR_WANT_ACCEPT"};
+			lwsl_info("%s: SSL_read: %s (%d); errno=%s (%d) (wsi=%p)\n",__func__, sslErrStr[err], err, strerror(errno_ssl_read), errno_ssl_read, wsi);
+		} else {
+			lwsl_err("%s failed1: %s\n",__func__,
+					ERR_error_string(err, NULL));
+			lws_decode_ssl_error();
+		}
+		return LWS_SSL_CAPABLE_ERROR;
+#endif
 	}
 
 	if (n < 0) {
@@ -326,12 +400,153 @@
 			lws_set_blocking_send(wsi);
 		return LWS_SSL_CAPABLE_MORE_SERVICE;
 	}
+	if (n > SSL_ERROR_NONE && n < SSL_ERROR_WANT_ACCEPT) {
+		return LWS_SSL_CAPABLE_ERROR;
+	} else {
 		lwsl_err("%s failed: %s\n",__func__,
-				 ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
-			lws_decode_ssl_error();
-
-
+				ERR_error_string(n, NULL));
+		lws_decode_ssl_error();
+	}
 	return LWS_SSL_CAPABLE_ERROR;
+}
+
+/*
+// precond: lws_is_ssl(wsi) > 0
+// return 1: success
+// retrun 0: failure, try again
+// return -1: failure, wsi == 0 || wsi->ssl == 0
+*/
+LWS_VISIBLE int
+lws_ssl_shutdown(struct lws *wsi)
+{
+#ifndef LWS_OPENSSL_SUPPORT
+	return 1;
+#endif
+
+	if (wsi == NULL || wsi->ssl == NULL) return -1;
+
+#ifndef _DEBUG
+	lws_rx_flow_control(wsi, 0);	/* stop rx inflow */
+	int shutdown_state = SSL_get_shutdown(wsi->ssl);
+
+	if (shutdown_state == 0) {	/* no shutdown sent or received yet; as initiator send a shutdown */
+		int ret = SSL_shutdown(wsi->ssl);	/* first shutdown, if successful it will return 0, in case of error -1 */
+		if (ret == 0) {
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, 0); 	/* reset state; note: even a failed SSL_shutdown (== -1) sets state to SSL_SENT_SHUTDOWN */
+		// if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 0 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			errno_shutdown = 0;
+			ret = SSL_shutdown(wsi->ssl);
+			errno_shutdown = errno;
+			sslErr = SSL_get_error(wsi->ssl, ret);
+		}
+		lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+		if (ret != 0) {
+			return 0;
+		} else {	// success
+			return 1;
+		}
+	} else if ((shutdown_state & SSL_RECEIVED_SHUTDOWN) && !(shutdown_state & SSL_SENT_SHUTDOWN)) {	/* just received (in SSL_read) a 'close notifiy' alert (=SSL_RECEIVED_SHUTDOWN) from our peer, but havent't sent yet a 'close notifiy' alert reply back (SSL_SENT_SHUTDOWN) */
+		int ret = SSL_shutdown(wsi->ssl);	/* success: 1; failure: -1; ignore outcome, peer could already teared down the socket; note: SSL_RECEIVED_SHUTDOWN already fulfills SSL/TLS protocol specification, 2nd step is optional (only mandatory if same ssl-socket-fd will be reused, what libwebsocket is not doing) */
+		errno_shutdown = errno;
+		if (ret == 1) {
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, SSL_RECEIVED_SHUTDOWN); 	/* set state back to SSL_RECEIVED_SHUTDOWN; note: even a failed SSL_shutdown (== -1) sets state SSL_SENT_SHUTDOWN */
+		/* if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent */
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 1 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			errno_shutdown = 0;
+			ret = SSL_shutdown(wsi->ssl);
+			errno_shutdown = errno;
+			sslErr = SSL_get_error(wsi->ssl, ret);
+		}
+		/* don't checking ret means the chance to silently accept a failed SSL_shutdonw here
+		 * but even with error: received 'close notifiy' alert from peer already fulfills SSL/TLS specification */
+	}
+	lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+	return 1;
+#else
+
+	lws_rx_flow_control(wsi, 0);	/* stop rx inflow */
+	int shutdown_state = SSL_get_shutdown(wsi->ssl);
+
+	int errno_shutdown = 0;
+	if (shutdown_state == 0) {	/* no shutdown sent or received yet; as initiator send a shutdown */
+		int ret = SSL_shutdown(wsi->ssl);	/* first shutdown, if successful it will return 0, in case of error -1 */
+		errno_shutdown = errno;
+		if (ret == 0) {
+			lwsl_info("%s: successfully sent initiating 'close notifiy' alert to peer (wsi=%p)\n",__func__, wsi);
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, 0); 	/* reset state; note: even a failed SSL_shutdown (== -1) sets state to SSL_SENT_SHUTDOWN */
+		print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		/* if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent */
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 0 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			errno_shutdown = 0;
+			ret = SSL_shutdown(wsi->ssl);
+			errno_shutdown = errno;
+			sslErr = SSL_get_error(wsi->ssl, ret);
+			print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		}
+		lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+		if (ret != 0) {
+			lwsl_info("%s: failed to send initiating 'close notifiy' alert to peer (wsi=%p)\n",__func__, wsi);
+			return 0;
+		} else {
+			lwsl_info("%s: successfully sent initiating 'close notifiy' alert to peer (wsi=%p)\n",__func__, wsi);
+			return 1;
+		}
+	} else if ((shutdown_state & SSL_RECEIVED_SHUTDOWN) && !(shutdown_state & SSL_SENT_SHUTDOWN)) {	/* just received (in SSL_read) a 'close notifiy' alert (=SSL_RECEIVED_SHUTDOWN) from our peer, but havent't sent yet a 'close notifiy' alert reply back (SSL_SENT_SHUTDOWN) */
+		lwsl_info("%s: (shutdown_state & SSL_RECEIVED_SHUTDOWN) && !(shutdown_state & SSL_SENT_SHUTDOWN) (wsi=%p)\n",__func__, wsi);
+		int ret = SSL_shutdown(wsi->ssl);	/* success: 1; failure: -1; ignore outcome, peer could already teared down the socket; note: SSL_RECEIVED_SHUTDOWN already fulfills SSL/TLS protocol specification, 2nd step is optional (only mandatory if same ssl-socket-fd will be reused, what libwebsocket is not doing) */
+		errno_shutdown = errno;
+		if (ret == 1) {
+			lwsl_info("%s: shutdown process successfully finished (wsi=%p)\n",__func__, wsi);
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, SSL_RECEIVED_SHUTDOWN); 	/* set state back to SSL_RECEIVED_SHUTDOWN; note: even a failed SSL_shutdown (== -1) sets state SSL_SENT_SHUTDOWN */
+		print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		/* if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent */
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 1 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			errno_shutdown = 0;
+			ret = SSL_shutdown(wsi->ssl);
+			errno_shutdown = errno;
+			sslErr = SSL_get_error(wsi->ssl, ret);
+			print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		}
+		if (ret == 1)
+			lwsl_info("%s: shutdown process successfully finished (wsi=%p)\n",__func__, wsi);
+		else
+			lwsl_info("%s: shutdown process successfully finished (even with error: received 'close notifiy' alert from peer already fulfills SSL/TLS specification) (wsi=%p)\n",__func__, wsi);
+	} else {
+		print_ssl_shutdown_state1(shutdown_state, wsi, __func__);
+		if ((shutdown_state & SSL_RECEIVED_SHUTDOWN) && (shutdown_state & SSL_SENT_SHUTDOWN))
+			lwsl_info("%s: nothing to do; shutdown process successfuly finished (wsi=%p)\n",__func__, wsi);
+		else if (!(shutdown_state & SSL_RECEIVED_SHUTDOWN) && (shutdown_state & SSL_SENT_SHUTDOWN))
+			lwsl_info("%s: no 'close notifiy' alert from peer yet; shutdown process still ongoing (note: successfully sent'close notifiy' alert to peer already fulfills SSL/TLS specification) (wsi=%p)\n",__func__, wsi);
+	}
+	lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+	return 1;
+
+#endif
 }
 
 LWS_VISIBLE int
@@ -342,8 +557,8 @@
 	if (!wsi->ssl)
 		return 0; /* not handled */
 
+	lws_ssl_shutdown(wsi);
 	n = SSL_get_fd(wsi->ssl);
-	SSL_shutdown(wsi->ssl);
 	compatible_close(n);
 	SSL_free(wsi->ssl);
 	wsi->ssl = NULL;
@@ -466,7 +681,7 @@
 				*/
 				wsi->use_ssl = 0;
 
-				SSL_shutdown(wsi->ssl);
+				lws_ssl_shutdown(wsi);
 				SSL_free(wsi->ssl);
 				wsi->ssl = NULL;
 				if (lws_check_opt(context->options,
```
#######################################################
```
diff --git a/lib/private-libwebsockets.h b/lib/private-libwebsockets.h
index 640e56a..0f30f43 100644
--- a/lib/private-libwebsockets.h
+++ b/lib/private-libwebsockets.h
@@ -1755,6 +1755,7 @@
 #define lws_ssl_capable_write lws_ssl_capable_write_no_ssl
 #define lws_ssl_pending lws_ssl_pending_no_ssl
 #define lws_server_socket_service_ssl(_b, _c) (0)
+#define lws_ssl_shutdown(_a) (0)
 #define lws_ssl_close(_a) (0)
 #define lws_ssl_context_destroy(_a)
 #define lws_ssl_SSL_CTX_destroy(_a)
@@ -1773,6 +1774,8 @@
 lws_context_init_ssl_library(struct lws_context_creation_info *info);
 LWS_EXTERN int LWS_WARN_UNUSED_RESULT
 lws_server_socket_service_ssl(struct lws *new_wsi, lws_sockfd_type accept_fd);
+LWS_VISIBLE int
+lws_ssl_shutdown(struct lws *wsi);
 LWS_EXTERN int
 lws_ssl_close(struct lws *wsi);
 LWS_EXTERN void
```

**Comment by namowen** [comment]
Correction, ssl.c didn't compile for release version. Here the correct patch for ssl.c:
```
diff --git a/lib/ssl.c b/lib/ssl.c
index 8c0d92b..705e107 100644
--- a/lib/ssl.c
+++ b/lib/ssl.c
@@ -86,6 +86,58 @@
 	}
 }
 
+#ifdef LWS_OPENSSL_SUPPORT
+
+#ifdef _DEBUG
+#if defined(LWS_WITH_NO_LOGS)
+#define print_ssl_shutdown_state1(a,b,c)
+#define print_ssl_shutdown_state2(a,b)
+#define print_ssl_shutdown_outcome(a,b,c,d)
+#else
+#define print_ssl_shutdown_state1(a,b,c) _print_ssl_shutdown_state1(a,b,c)
+#define print_ssl_shutdown_state2(a,b) _print_ssl_shutdown_state2(a,b)
+#define print_ssl_shutdown_outcome(a,b,c,d) _print_ssl_shutdown_outcome(a,b,c,d)
+#endif
+#endif
+
+#ifdef _DEBUG
+
+void _print_ssl_shutdown_state1(int shutdown_state, const struct lws *wsi, const char funcName[]) {
+	if (shutdown_state & SSL_RECEIVED_SHUTDOWN) {
+		if(shutdown_state & SSL_SENT_SHUTDOWN) {
+			lwsl_info("%s SSL_get_shutdown (%d): SSL_RECEIVED_SHUTDOWN & SSL_SENT_SHUTDOWN (wsi=%p)\n",funcName, shutdown_state, wsi);
+		}
+		else lwsl_info("%s SSL_get_shutdown (%d): SSL_RECEIVED_SHUTDOWN (wsi=%p)\n",funcName, shutdown_state, wsi);
+	} else if(shutdown_state == 0) {
+		lwsl_info("%s SSL_get_shutdown (%d): No shutdown setting, yet. (wsi=%p)\n",funcName, shutdown_state, wsi);
+	} else if(shutdown_state  & SSL_SENT_SHUTDOWN) {
+		lwsl_info("%s SSL_get_shutdown (%d): SSL_SENT_SHUTDOWN (wsi=%p)\n",funcName, shutdown_state, wsi);
+	}
+}
+
+void _print_ssl_shutdown_state2(const struct lws *wsi, const char funcName[]) {
+	int shutdown_state = SSL_get_shutdown(wsi->ssl);
+	_print_ssl_shutdown_state1(shutdown_state, wsi, funcName);
+}
+
+void _print_ssl_shutdown_outcome(int outcome, int shutdown_errno, const struct lws *wsi, const char funcName[]) {
+	if (outcome == 0) lwsl_info("%s: SSL_shutdown (%d) notify alert sent successfully, awaiting response from peer (wsi=%p)\n",funcName, outcome, wsi);
+	if (outcome == 1) lwsl_info("%s: 2nd SSL_shutdown (%d) notify alert sent successfully, after receiving notify alert response from peer (wsi=%p)\n",funcName, outcome, wsi);
+	if (outcome < 0) {
+		lwsl_info("%s: SSL_shutdown=%d error (wsi=%p)\n",funcName, outcome, wsi);
+		int sslErr = SSL_get_error(wsi->ssl, outcome);
+		if (sslErr > SSL_ERROR_NONE && sslErr < SSL_ERROR_WANT_ACCEPT) {
+			char *sslErrStr[] = {"SSL_ERROR_NONE","SSL_ERROR_SSL","SSL_ERROR_WANT_READ","SSL_ERROR_WANT_WRITE","SSL_ERROR_WANT_X509_LOOKUP","SSL_ERROR_SYSCALL","SSL_ERROR_ZERO_RETURN","SSL_ERROR_WANT_CONNECT","SSL_ERROR_WANT_ACCEPT"};
+			lwsl_info("%s: SSL_shutdown: %s (%d); errno=%s (%d) (wsi=%p)\n",funcName, sslErrStr[sslErr], sslErr, strerror(shutdown_errno), shutdown_errno, wsi);
+		} else {
+			lwsl_err("%s failed: %s\n",funcName,	ERR_error_string(sslErr, NULL));
+			lws_decode_ssl_error();
+		}
+	}
+}
+#endif
+#endif
+
 static int
 lws_context_init_ssl_pem_passwd_cb(char * buf, int size, int rwflag, void *userdata)
 {
@@ -237,12 +289,34 @@
 	n = SSL_read(wsi->ssl, buf, len);
 
 	/* manpage: returning 0 means connection shut down */
-	if (!n) {
-		lwsl_err("%s failed: %s\n",__func__,
-			 ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
-		lws_decode_ssl_error();
-
+	if (n == 0) {
+#ifndef _DEBUG
+		int err = SSL_get_error(wsi->ssl, n);
+		if (err > SSL_ERROR_NONE && err < SSL_ERROR_WANT_ACCEPT) {
+			return LWS_SSL_CAPABLE_ERROR;
+		} else {
+			lwsl_err("%s failed: %s\n",__func__,
+					ERR_error_string(err, NULL));
+			lws_decode_ssl_error();
+		}
 		return LWS_SSL_CAPABLE_ERROR;
+#else
+		int errno_ssl_read = errno;
+		int err = SSL_get_error(wsi->ssl, n);
+		if (err == SSL_ERROR_ZERO_RETURN) {
+			print_ssl_shutdown_state2(wsi, __func__);
+			lwsl_info("%s: SSL_read: SSL_ERROR_ZERO_RETURN (%d); (wsi=%p)\n",__func__, SSL_ERROR_ZERO_RETURN, wsi);
+			return LWS_SSL_CAPABLE_ERROR;
+		} else if (err > SSL_ERROR_NONE && err < SSL_ERROR_WANT_ACCEPT) {
+			char *sslErrStr[] = {"SSL_ERROR_NONE","SSL_ERROR_SSL","SSL_ERROR_WANT_READ","SSL_ERROR_WANT_WRITE","SSL_ERROR_WANT_X509_LOOKUP","SSL_ERROR_SYSCALL","SSL_ERROR_ZERO_RETURN","SSL_ERROR_WANT_CONNECT","SSL_ERROR_WANT_ACCEPT"};
+			lwsl_info("%s: SSL_read: %s (%d); errno=%s (%d) (wsi=%p)\n",__func__, sslErrStr[err], err, strerror(errno_ssl_read), errno_ssl_read, wsi);
+		} else {
+			lwsl_err("%s failed1: %s\n",__func__,
+					ERR_error_string(err, NULL));
+			lws_decode_ssl_error();
+		}
+		return LWS_SSL_CAPABLE_ERROR;
+#endif
 	}
 
 	if (n < 0) {
@@ -326,12 +400,148 @@
 			lws_set_blocking_send(wsi);
 		return LWS_SSL_CAPABLE_MORE_SERVICE;
 	}
+	if (n > SSL_ERROR_NONE && n < SSL_ERROR_WANT_ACCEPT) {
+		return LWS_SSL_CAPABLE_ERROR;
+	} else {
 		lwsl_err("%s failed: %s\n",__func__,
-				 ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
-			lws_decode_ssl_error();
-
-
+				ERR_error_string(n, NULL));
+		lws_decode_ssl_error();
+	}
 	return LWS_SSL_CAPABLE_ERROR;
+}
+
+/*
+// precond: lws_is_ssl(wsi) > 0
+// return 1: success
+// retrun 0: failure, try again
+// return -1: failure, wsi == 0 || wsi->ssl == 0
+*/
+LWS_VISIBLE int
+lws_ssl_shutdown(struct lws *wsi)
+{
+#ifndef LWS_OPENSSL_SUPPORT
+	return 1;
+#endif
+
+	if (wsi == NULL || wsi->ssl == NULL) return -1;
+
+#ifndef _DEBUG
+	lws_rx_flow_control(wsi, 0);	/* stop rx inflow */
+	int shutdown_state = SSL_get_shutdown(wsi->ssl);
+
+	if (shutdown_state == 0) {	/* no shutdown sent or received yet; as initiator send a shutdown */
+		int ret = SSL_shutdown(wsi->ssl);	/* first shutdown, if successful it will return 0, in case of error -1 */
+		if (ret == 0) {
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, 0); 	/* reset state; note: even a failed SSL_shutdown (== -1) sets state to SSL_SENT_SHUTDOWN */
+		// if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 0 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			ret = SSL_shutdown(wsi->ssl);
+			sslErr = SSL_get_error(wsi->ssl, ret);
+		}
+		lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+		if (ret != 0) {
+			return 0;
+		} else {	// success
+			return 1;
+		}
+	} else if ((shutdown_state & SSL_RECEIVED_SHUTDOWN) && !(shutdown_state & SSL_SENT_SHUTDOWN)) {	/* just received (in SSL_read) a 'close notifiy' alert (=SSL_RECEIVED_SHUTDOWN) from our peer, but havent't sent yet a 'close notifiy' alert reply back (SSL_SENT_SHUTDOWN) */
+		int ret = SSL_shutdown(wsi->ssl);	/* success: 1; failure: -1; ignore outcome, peer could already teared down the socket; note: SSL_RECEIVED_SHUTDOWN already fulfills SSL/TLS protocol specification, 2nd step is optional (only mandatory if same ssl-socket-fd will be reused, what libwebsocket is not doing) */
+		if (ret == 1) {
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, SSL_RECEIVED_SHUTDOWN); 	/* set state back to SSL_RECEIVED_SHUTDOWN; note: even a failed SSL_shutdown (== -1) sets state SSL_SENT_SHUTDOWN */
+		/* if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent */
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 1 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			ret = SSL_shutdown(wsi->ssl);
+			sslErr = SSL_get_error(wsi->ssl, ret);
+		}
+		/* don't checking ret means the chance to silently accept a failed SSL_shutdonw here
+		 * but even with error: received 'close notifiy' alert from peer already fulfills SSL/TLS specification */
+	}
+	lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+	return 1;
+#else
+
+	lws_rx_flow_control(wsi, 0);	/* stop rx inflow */
+	int shutdown_state = SSL_get_shutdown(wsi->ssl);
+
+	int errno_shutdown = 0;
+	if (shutdown_state == 0) {	/* no shutdown sent or received yet; as initiator send a shutdown */
+		int ret = SSL_shutdown(wsi->ssl);	/* first shutdown, if successful it will return 0, in case of error -1 */
+		errno_shutdown = errno;
+		if (ret == 0) {
+			lwsl_info("%s: successfully sent initiating 'close notifiy' alert to peer (wsi=%p)\n",__func__, wsi);
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, 0); 	/* reset state; note: even a failed SSL_shutdown (== -1) sets state to SSL_SENT_SHUTDOWN */
+		print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		/* if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent */
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 0 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			errno_shutdown = 0;
+			ret = SSL_shutdown(wsi->ssl);
+			errno_shutdown = errno;
+			sslErr = SSL_get_error(wsi->ssl, ret);
+			print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		}
+		lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+		if (ret != 0) {
+			lwsl_info("%s: failed to send initiating 'close notifiy' alert to peer (wsi=%p)\n",__func__, wsi);
+			return 0;
+		} else {
+			lwsl_info("%s: successfully sent initiating 'close notifiy' alert to peer (wsi=%p)\n",__func__, wsi);
+			return 1;
+		}
+	} else if ((shutdown_state & SSL_RECEIVED_SHUTDOWN) && !(shutdown_state & SSL_SENT_SHUTDOWN)) {	/* just received (in SSL_read) a 'close notifiy' alert (=SSL_RECEIVED_SHUTDOWN) from our peer, but havent't sent yet a 'close notifiy' alert reply back (SSL_SENT_SHUTDOWN) */
+		lwsl_info("%s: (shutdown_state & SSL_RECEIVED_SHUTDOWN) && !(shutdown_state & SSL_SENT_SHUTDOWN) (wsi=%p)\n",__func__, wsi);
+		int ret = SSL_shutdown(wsi->ssl);	/* success: 1; failure: -1; ignore outcome, peer could already teared down the socket; note: SSL_RECEIVED_SHUTDOWN already fulfills SSL/TLS protocol specification, 2nd step is optional (only mandatory if same ssl-socket-fd will be reused, what libwebsocket is not doing) */
+		errno_shutdown = errno;
+		if (ret == 1) {
+			lwsl_info("%s: shutdown process successfully finished (wsi=%p)\n",__func__, wsi);
+			lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+			return 1;
+		}
+		SSL_set_shutdown(wsi->ssl, SSL_RECEIVED_SHUTDOWN); 	/* set state back to SSL_RECEIVED_SHUTDOWN; note: even a failed SSL_shutdown (== -1) sets state SSL_SENT_SHUTDOWN */
+		print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		/* if SSL_ERROR_WANT_WRITE: retry 5 times every second again to sent */
+		int repeat = 5;
+		int sslErr = SSL_get_error(wsi->ssl, ret);
+		while (ret != 1 && sslErr == SSL_ERROR_WANT_WRITE && repeat--) {
+			sleep(1);
+			errno_shutdown = 0;
+			ret = SSL_shutdown(wsi->ssl);
+			errno_shutdown = errno;
+			sslErr = SSL_get_error(wsi->ssl, ret);
+			print_ssl_shutdown_outcome(ret, errno_shutdown, wsi,__func__);
+		}
+		if (ret == 1)
+			lwsl_info("%s: shutdown process successfully finished (wsi=%p)\n",__func__, wsi);
+		else
+			lwsl_info("%s: shutdown process successfully finished (even with error: received 'close notifiy' alert from peer already fulfills SSL/TLS specification) (wsi=%p)\n",__func__, wsi);
+	} else {
+		print_ssl_shutdown_state1(shutdown_state, wsi, __func__);
+		if ((shutdown_state & SSL_RECEIVED_SHUTDOWN) && (shutdown_state & SSL_SENT_SHUTDOWN))
+			lwsl_info("%s: nothing to do; shutdown process successfuly finished (wsi=%p)\n",__func__, wsi);
+		else if (!(shutdown_state & SSL_RECEIVED_SHUTDOWN) && (shutdown_state & SSL_SENT_SHUTDOWN))
+			lwsl_info("%s: no 'close notifiy' alert from peer yet; shutdown process still ongoing (note: successfully sent'close notifiy' alert to peer already fulfills SSL/TLS specification) (wsi=%p)\n",__func__, wsi);
+	}
+	lws_rx_flow_control(wsi, 1);	/* enable rx inflow */
+	return 1;
+
+#endif
 }
 
 LWS_VISIBLE int
@@ -342,8 +552,8 @@
 	if (!wsi->ssl)
 		return 0; /* not handled */
 
+	lws_ssl_shutdown(wsi);
 	n = SSL_get_fd(wsi->ssl);
-	SSL_shutdown(wsi->ssl);
 	compatible_close(n);
 	SSL_free(wsi->ssl);
 	wsi->ssl = NULL;
@@ -466,7 +676,7 @@
 				*/
 				wsi->use_ssl = 0;
 
-				SSL_shutdown(wsi->ssl);
+				lws_ssl_shutdown(wsi);
 				SSL_free(wsi->ssl);
 				wsi->ssl = NULL;
 				if (lws_check_opt(context->options,

```

**Comment by namowen** [comment]
With my extended tests (instead localhost to localhost I test Android smartphone via WLAN to server), I've got a difficult reproducable 'read failed' on mirror-protocol in the client. This comes out of LWS_SSL_CAPABLE_ERROR after lws_ssl_capable_read.
I have to investigate why this is happening (loglevel was only 1...). Could be, that Resource temporarily unavailable...

**Comment by lws-team** [comment]
Thanks... I agree this is difficult, but lws cannot completely block for up to 5s per ssl connection close to solve it.  Lws may be handling many thousands of concurrent server and / or client connections, ssl close needs to be dealt with also in nonblocking fashion.

From lws perspective, by comparison the current EVP "errors" under the condition we hang up the connection are symptomless and much preferable.

Basically we deal with WANT_WRITE/READ already under normal operation, it has to be handled by returning to the event loop and coming back for retry when it's our turn.  I understand you "solved your [current] problem" with this though and thanks for sharing.

**Comment by lws-team** [comment]
I think we hit the end of the original subject for the moment, since we are coming up to a release and the plan is leave the EVP thing until later.  lws_ssl_capable_read() handles WANT_WRITE/READ.  If your Android problem turns out to involve lws code please open another issue.

**Comment by namowen** [comment]
I agree. In a single-threaded environment handling thousands of connections, this patch is not appropriate.
Here is a far simpler solution, handling DH lib and EVP lib errors more appropriate.
```
diff --git a/lib/ssl.c b/lib/ssl.c
index 8c0d92b..0608e9e 100644
--- a/lib/ssl.c
+++ b/lib/ssl.c
@@ -230,6 +230,7 @@
 	struct lws_context *context = wsi->context;
 	struct lws_context_per_thread *pt = &context->pt[(int)wsi->tsi];
 	int n = 0;
+	int ssl_read_errno = 0;
 
 	if (!wsi->ssl)
 		return lws_ssl_capable_read_no_ssl(wsi, buf, len);
@@ -238,6 +239,20 @@
 
 	/* manpage: returning 0 means connection shut down */
 	if (!n) {
+		n = lws_ssl_get_error(wsi, n);
+
+		if (n == SSL_ERROR_ZERO_RETURN)
+			return LWS_SSL_CAPABLE_ERROR;
+
+		if (n == SSL_ERROR_SYSCALL) {
+			int err = ERR_get_error();
+			if (err == 0
+					&& (ssl_read_errno == EPIPE
+						|| ssl_read_errno == ECONNABORTED
+						|| ssl_read_errno == 0))
+					return LWS_SSL_CAPABLE_ERROR;
+		}
+
 		lwsl_err("%s failed: %s\n",__func__,
 			 ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
 		lws_decode_ssl_error();
@@ -312,6 +327,7 @@
 lws_ssl_capable_write(struct lws *wsi, unsigned char *buf, int len)
 {
 	int n;
+	int ssl_read_errno = 0;
 
 	if (!wsi->ssl)
 		return lws_ssl_capable_write_no_ssl(wsi, buf, len);
@@ -326,10 +342,22 @@
 			lws_set_blocking_send(wsi);
 		return LWS_SSL_CAPABLE_MORE_SERVICE;
 	}
-		lwsl_err("%s failed: %s\n",__func__,
-				 ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
-			lws_decode_ssl_error();
 
+	if (n == SSL_ERROR_ZERO_RETURN)
+		return LWS_SSL_CAPABLE_ERROR;
+
+	if (n == SSL_ERROR_SYSCALL) {
+		int err = ERR_get_error();
+		if (err == 0
+				&& (ssl_read_errno == EPIPE
+					|| ssl_read_errno == ECONNABORTED
+					|| ssl_read_errno == 0))
+				return LWS_SSL_CAPABLE_ERROR;
+	}
+
+	lwsl_err("%s failed: %s\n",__func__,
+			ERR_error_string(lws_ssl_get_error(wsi, 0), NULL));
+	lws_decode_ssl_error();
 
 	return LWS_SSL_CAPABLE_ERROR;
 }

```

**Comment by lws-team** [comment]
Thanks... that is a lot better for general use... I confirmed the EVP logging is gone and pushed it on master.  Thanks a lot for having another crack at it.
